### PR TITLE
[Flang][OpenMP][MLIR] Initial derived type member map support

### PR DESCRIPTION
**Pull request by agozillon**
This patch is one in a stack of four patches that seeks to refactor
slightly and extend the current record type map support that was
put in place for Fortran's descriptor types to handle explicit
member mapping for record types at a single level of depth.

For example, the below case where members of Fortran
derived types are mapped explicitly:

```
  type :: scalar_and_array
    real(4) :: real
    integer(4) :: array(10)
    integer(4) :: int
  end type scalar_and_array
  type(scalar_and_array) :: scalar_arr

  !$omp target map(tofrom: scalar_arr%int, scalar_arr%real)
```
```
    type :: bottom_layer
      real(8) :: i2
      real(4) :: array_i2(10)
      real(4) :: array_j2(10)
    end type bottom_layer

    type :: top_layer
      real(4) :: i
      integer(4) :: array_i(10)
      real(4) :: j
      type(bottom_layer) :: nested
      integer, allocatable :: array_j(:)
      integer(4) :: k
    end type top_layer
    
    type(top_layer) :: top_dtype

!$omp target map(tofrom: top_dtype%nested%i2, top_dtype%k, top_dtype%nested%array_i2)
```
Current cases of derived type mapping left for future work are:

-   Fortran's automagical mapping of all elements and nested elements of a derived type
-   explicit member mapping of a derived type and then constituent members (redundant in Fortran due to former case but still legal as far as I am aware)
-   explicit member mapping of a record type (may be handled reasonably, just not fully tested in this iteration)
-   explicit member mapping for Fortran allocatable types (a variation of nested record types)

This patch seeks to support this by extending the Flang-new OpenMP lowering to
support generation of this newly required information, creating the necessary
parent <-to-> member map_info links, calculating the member indices and
setting if it's a partial map.

The OMPDescriptorMapInfoGen pass has also been generalized into a map
finalization phase, now named OMPMapInfoFinalization. This pass was extended
to support the insertion of member maps into the BlockArg and MapOperands of
relevant map carrying operations. Similar to the method in which descriptor types
are expanded and constituent members inserted.

### CODE REVIEW

**Comment by agozillon** [comment]
> You could do the renaming of the pass as an NFC PR so it would make the non NFC changes more obvious to spot at. Not necessary to do it for this time but next time it would be nice.

I'm very sorry about that! I didn't realise it would be a problem, but I will keep it in mind for the future for any renaming + change sets I aim to upstream. Thank you for bringing it up.

**Comment by clementval** [comment]
> > You could do the renaming of the pass as an NFC PR so it would make the non NFC changes more obvious to spot at. Not necessary to do it for this time but next time it would be nice.
> 
> I'm very sorry about that! I didn't realise it would be a problem, but I will keep it in mind for the future for any renaming + change sets I aim to upstream. Thank you for bringing it up.

Not an issue at all. It's just simpler to look at small PRs :-)

**Comment by agozillon** [comment]
I have opened PRs again on the entire stack, hopefully they shouldn't cause problems for windows users anymore, I've checked the length of the largest PR and I believe it fits within the maximum that we're planning on enforcing.

**Comment by agozillon** [comment]
> > > You could do the renaming of the pass as an NFC PR so it would make the non NFC changes more obvious to spot at. Not necessary to do it for this time but next time it would be nice.
> > 
> > 
> > I'm very sorry about that! I didn't realise it would be a problem, but I will keep it in mind for the future for any renaming + change sets I aim to upstream. Thank you for bringing it up.
> 
> Not an issue at all. It's just simpler to look at small PRs :-)

Yes, I am sorry about that, I'd like to make it as easy as possible to review for you all as I know it's not an easy task and I greatly appreciate the effort you all put in, so I will most definitely keep it mind in the future!

**Comment by clementval** [comment]
> Yes, I am sorry about that, I'd like to make it as easy as possible to review for you all as I know it's not an easy task and I greatly appreciate the effort you all put in, so I will most definitely keep it mind in the future!

Thanks!

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp
Comment (kparzysz) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:941]: We will need to traverse all parents.  For example given `x%y%z`, `GetFirstName` will return `x`.  The type of `x` does not have a member `z`, so findComponentMemberPlacement will fail.

Comment (agozillon) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:941]: This PR isn't intended to handle that type of mapping yet, it'll be handled in a subsequent series (which I am working on at the moment). 

It's only intended to handle cases where we're mapping one level of nesting for the moment.

**Comment by agozillon** [comment]
The current PR stack has been updated with the aims to extend support to explicit derived type member mapping at arbitrary depths, and not just top level member mapping. The change set is slightly different, primarily in the sense that we now need to handle index generation in Flang and then first/last member selection in the lowering differently due to the new attribute type in use for member placement tracking (DenseIntElementsAttr, for N-D indices). 

These are the encompassing PR stack that were updated:

https://github.com/llvm/llvm-project/pull/82853
https://github.com/llvm/llvm-project/pull/82852
https://github.com/llvm/llvm-project/pull/82851
https://github.com/llvm/llvm-project/pull/82850

The only PR from the stack that is still a little in flux so I would recommend not reviewing for a little while (I'll add a comment when it's ready), is the MLIR -> LLVM-IR lowering (https://github.com/llvm/llvm-project/pull/82852), as it has a partially duplicate change set that's segmented into another PR (https://github.com/llvm/llvm-project/pull/84349) for hopefully easier/quicker review. Once the other has landed I hope to perform a rebase to align the stack completely. That being said, if you do review it, I'll do my best to apply the requests to the relevant PR.

**Comment by agozillon** [comment]
Rebased the entire stack on recent upstream changes, there was quite a few changes that meant some sizeable alterations to the Fortran lowering segment of the PR stack, but otherwise it all went smoothly. However, please do point out if you see anything out of the ordinary across the change set that may look like a rebase artifact when reviewing!

Otherwise, this PR stack is now fully ready for review so please do so if you have some time, it would be greatly appreciated.

**Comment by agozillon** [comment]
Small ping for review of this PR stack if possible. I'll be away on vacation for two and a half weeks, but it would be excellent to come back to review comments on this PR stack that I can address and then make some forward progress on. No rush obviously as there will be a 2 week time lag before it's all addressed but reviews would be appreciated!

**Comment by agozillon** [comment]
Back from vacation, would very much love if a kind reviewer or two could please allocate some time to review this PR stack so I can hopefully move it further along in the next few weeks :-) Thank you ahead of time!

File: flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp:174]:
```suggestion
  // The parent member becomes more critical, as we perform a partial
  // structure mapping where we link the mapping of x and y together
  // via the parent. We do this at a kernel argument level in LLVM IR and not just
  // MLIR, which is important to maintain similarity to Clang and for the runtime
  // to do the correct thing. However, we still do not map the structure
  // in its totality but rather we generate an un-sized "binding"
  // map entry for it.
```

File: flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp:181]: Nit: I'd prefer saying that it's redundant rather than saying that it doesn't make sense to do, but at the end of the day it's the same thing, so feel free to ignore if you disagree.

File: flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp:201]:
```suggestion
        for (auto [j, mapMember] : llvm::enumerate(op.getMembers())) {
```

File: flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp:210]: Nit: Adding an unknown location will result in unhelpful messages if there's a verifier problem later on. I'm not sure if in this case it's better to use the location of the `MapInfoOp` or the `TargetOp`, but I believe either one will be better than nothing.
```suggestion
                i + j, op.getMembers()[j].getType(), op.getLoc());
```

File: flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp
Comment (skatrak) [inline flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp:57]:
```suggestion
    mlir::Location loc = op.getLoc();
```

File: flang/test/Lower/OpenMP/FIR/map-component-ref.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/FIR/map-component-ref.f90:22]: Nit: Accidental change?

**Review by skatrak — COMMENTED** [review]
Thank you for all the work here Andrew, I've got some comments but they should be relatively easy to address.

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp
Comment (agozillon) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:811]: I can move it :-) the implementation just seems to have missed the migration as the declaration exists in the Utils header.

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp
Comment (agozillon) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:920]: It's expected behavior as there's nothing we can really do without the information, we'd be missing the parent symbol required to perform the next steps unfortunately. So I'll emit an assert in this case, as there likely is no fallback (at least that I can think of).

File: flang/lib/Lower/OpenMP/ClauseProcessor.h
Comment (agozillon) [reply flang/lib/Lower/OpenMP/ClauseProcessor.h:184]: I agree and have been thinking about it, as there isn't a whole lot of dissimilarities, most changes to processMapClauses need replicated to processMotionClauses and it would make testing a lot simpler! But it'd also be up to @ergawy I am unsure if he'd like to keep them distinct and has some insights into motion clauses that might make merging the two functions unideal! 

But I can make a follow up PR after the stack has landed to merge them if we are all happy doing so.

File: flang/lib/Lower/OpenMP/ClauseProcessor.h
Comment (agozillon) [reply flang/lib/Lower/OpenMP/ClauseProcessor.h:118]: I had this originally, it felt a little messy to me from what I recall so I opted for this! But if you prefer the other option and it makes more sense to you, I am perfectly happy with it as well :-)

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (agozillon) [reply flang/lib/Lower/OpenMP/Utils.cpp:209]: I like the idea of more information in my asserts, but happy to swap to it :-)

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (agozillon) [reply flang/lib/Lower/OpenMP/Utils.cpp:272]: From my understanding (and someone do please chime in if I am incorrect or misunderstanding), it will get the far right symbol, whereas previously it would get the far left symbol. 

So given the following:

dtype1%dtype2%array[2:10]

The getFirstName method would have retrieved "dtype1", not ideal as that's not what the ompObject is really referencing or what we care about in the case of unwrapping an array as it isn't the arrays symbol.

In the case of getLastName, we'd end up with "array", which is more what we'd expect.  

As we never really used constructs like this with getOmpObjectSymbol before (from what I can tell at least) it wasn't noticeable I believe. As for array[2:10] either getFirstName or getLastName will resolve to "array".

File: flang/lib/Lower/OpenMP/Utils.h
Comment (agozillon) [reply flang/lib/Lower/OpenMP/Utils.h:27]: It is no longer used! Thank you for pointing that out.

File: flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp
Comment (agozillon) [reply flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp:181]: works for me, not overly bothered about the wording as long as it's understood :-)

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (agozillon) [reply flang/lib/Lower/OpenMP/Utils.cpp:148]: I've opted to not change these to llvm::ArrayRef's but rather llvm::SmallVectorImpl's, as ArrayRef doesn't seem to play particularly nicely with the resize and fill.

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (agozillon) [reply flang/lib/Lower/OpenMP/Utils.cpp:132]: Opted to take it in as an argument as opposed to a return type, as returning a local SmallVector to a SmallVectorImpl doesn't seem like it's allowed (perhaps I tried it incorrectly however, but I couldn't grep a case where it seemed done in LLVM, my grep capabilities are possibly lacking though)

**Comment by agozillon** [comment]
Performed a rebase across the entire PR stack (my appologies, I forgot that might make the review a little more irritating before I performed it), and made an attempt to address all reviewer comments and where things didn't quite work as expected left replies on the comments.

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (jsjodin) [inline flang/lib/Lower/OpenMP/Utils.cpp:258]: Nit: I think explicit types should be used if the type isn't specified on the rhs.

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (jsjodin) [inline flang/lib/Lower/OpenMP/Utils.cpp:261]: Explicit type here also.

**Review by jsjodin — COMMENTED** [review]
Looks good, just some minor nits/comments.

**Comment by agozillon** [comment]
Thanks for the review @jsjodin and @skatrak I've attempted to address both of your comments in recent updates.

**Review by jsjodin — APPROVED** [review]
LGTM

File: flang/lib/Lower/OpenMP/ClauseProcessor.h
Comment (skatrak) [reply flang/lib/Lower/OpenMP/ClauseProcessor.h]: I agree it doesn't look great, but in my opinion it's better to do this than to force the caller to pass an output list even when it won't use it because it's needed internally. Maybe we'll be able to improve things later as we keep developing this.

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP/Utils.cpp]: Yeah, I haven't seen that done often either. The comment was mainly to avoid hardcoding some default `SmallVector` size in the public interface of the function (in this case, the default one for `int`), so passing a reference works for me too.

**Review by skatrak — APPROVED** [review]
Thank you Andrew, you've addressed all my concerns so this LGTM.

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP/Utils.cpp]: You're right, elements in `llvm::ArrayRef` are read-only. Sorry about that, I hope I didn't make you waste too much time trying this!

File: flang/lib/Lower/OpenMP/Utils.cpp
Comment (skatrak) [reply flang/lib/Lower/OpenMP/Utils.cpp]: That's a good point. If you have a more informative message for this case, I'm fine with you putting that in a separate assert.

**Comment by agozillon** [comment]
Thank you all very much for the reviews across the PR stack! 

I believe I have enough acceptance for the PR stack to land now, with at least two reviewers per PR. So I will seek to land this PR stack in the next few days (on Thursday) I'll perform the final adjustments requested, rebase and test prior to landing.

However, if anyone has any concerns or further review comments for any segments, please mention them now and I can hold off the PR until they are addressed if necessary :-)

**Comment by agozillon** [comment]
I made a bit of a mistake not using SPR to land this PR series, so i've ended up having to directly commit the PR, the associated commit for this PR can be found here: https://github.com/llvm/llvm-project/commit/435e850ba97ab567a14b6c84d2b27cadb771cb27

**Pull request merged**
